jobs/audit_jobs.py

```diff
@@ -89,13 +123,53 @@ def run(self):
 
         model_groups = python_utils.ZIP(KIND_BY_INDEX, models_of_kind_by_index)
         for kind, models_of_kind in model_groups:
-            # NOTE: Using extend() instead of append() because ApplyAuditDoFns
-            # produces an iterable of PCollections rather than a single one.
-            # NOTE: Label is missing because ApplyAuditDoFns labels itself.
             audit_error_pcolls.extend(models_of_kind | ApplyAuditDoFns(kind))
 
+            if kind in ALL_MODEL_KINDS_REFERENCED_BY_PROPERTIES:
+                existing_key_count_pcolls.append(
+                    models_of_kind | GetExistingModelKeyCounts(kind))
+
+            if kind in ID_REFERENCING_PROPERTIES_BY_KIND_OF_POSSESSOR:
+                missing_key_error_pcolls.extend(
+                    models_of_kind | GetMissingModelKeyErrors(kind))
+
+        existing_key_counts = (
+            existing_key_count_pcolls
+            | 'Flatten PCollections of existing key counts' >> beam.Flatten()
+        )
+        missing_key_errors = (
+            missing_key_error_pcolls
+            | 'Flatten PCollections of missing key errors' >> beam.Flatten()
+        )
+        audit_error_pcolls.append(
+            (existing_key_counts, missing_key_errors)
+            | 'Group counts and errors by key' >> beam.CoGroupByKey()
+            | 'Filter keys without any errors' >> (
+                beam.FlatMapTuple(self._get_model_relationship_errors))
+        )
```

Comment:
If I understand this correctly we retrieve the counts for each possible property field and models that it refers to and also retrieve all possible errors that might occur. Then only if the model count is zero we show the errors.

Will this work if we need to only run the prod validation on a subset of models? Some of the models mentioned in the relationships won't be part of that subset.

Also, we in the legacy prod validators sometimes not only verify whether the models in relationship exist but also if they conform to some rule, is it possible to extend this code to do it?

Reply:
> If I understand this correctly we retrieve the counts for each possible property field and models that it refers to and also retrieve all possible errors that might occur. Then only if the model count is zero we show the errors.

Yup that's exactly right

> Will this work if we need to only run the prod validation on a subset of models? Some of the models mentioned in the relationships won't be part of that subset.

The code is being written to target all models to start with, so I don't think this will be a problem. We can limit it to a subset later, depending on the first run.

> Also, we in the legacy prod validators sometimes not only verify whether the models in relationship exist but also if they conform to some rule, is it possible to extend this code to do it?

It should be possible to extend `_get_model_relationship_errors` to perform more checks, yes. It'd probably be ideal to add a mapping `(Model, Model) -> DoFn` to handle those checks efficiently.